### PR TITLE
Add FilesystemLocator

### LINKED ISSUE #47: Add FilesystemLocator
Implement a generic FilesystemLocator and refactor the translator and template locators to use it.

**Breaking change**: Resource paths are now url's and need to be parseable using parse_url. 

- Fully qualified translation paths are formatted as : com:foo/en_US.[yaml]
- Fully qualified template paths are formatted as : com:foo/view/layout.format[.engine]

Fully qualified template paths used in import() methods need to be updated to follow the new format.

### PR DESCRIPTION
**Pull request by johanjanssens**
resolve #47

### CODE REVIEW

**Comment by johanjanssens** [comment]
@ercanozkaya Can you give this a thumbs up ?

File: code/filesystem/locator/abstract.php [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+        if (strpos($path, '://') === false)
+        {
+            // Not a stream, so do a realpath() to avoid directory traversal attempts on the local file system.
+            $path = realpath($path); // needed for substr() later
+            $file = realpath($file);
+        }
+
+        // The substr() check added to make sure that the realpath() results in a directory registered so that
+        // non-registered directories are not accessible via directory traversal attempts.
+        if (file_exists($file) && substr($file, 0, strlen($path)) == $path) {
```
Comment (oligriffiths) [inline code/filesystem/locator/abstract.php]: `=== $path` ?

File: code/filesystem/locator/component.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+     * The locator name
+     *
+     * @var string
+     */
+    protected static $_name = 'com';
+
+    /**
+     * Register a path template
+     *
+     * If the template path has format com:foo]/path/to/file this method will replace com:foo with the
```
Comment (oligriffiths) [inline code/filesystem/locator/component.php]: `com:foo]`?

File: code/filesystem/locator/factory.php [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        //If a windows drive letter is passed use file:// locator
+        if(strtoupper(substr(PHP_OS, 0, 3)) === 'WIN')
+        {
+            if(preg_match('#^[a-z]{1}$#i', $name)) {
+                $name = 'file';
+            }
+        }
+
+        //Locator not supported
+        if(!in_array($name, $this->getLocators()))
```
Comment (oligriffiths) [inline code/filesystem/locator/factory.php]: Should this not be an `isset()`? Looks like `$this-> __locators` is an array indexed by name

File: code/filesystem/locator/abstract.php [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        $path = str_replace(array($scheme.'://', $scheme.':'), '', dirname($url));
+        $parts = explode('/', $path);
+
+        $info = array(
+            'type'    => $type,
+            'domain'  => $domain,
+            'package' => array_shift($parts),
+            'path'    => implode('/', $parts),
+            'file'    => $file,
+            'format'  => $format ?: '*',
```
Comment (oligriffiths) [inline code/filesystem/locator/abstract.php]: Isnt `$format` always defined? Seems like it's either the penultimate or last item in the `$parts` array

**Comment by ercanozkaya** [comment]
@johanjanssens Looking good.

File: code/filesystem/locator/abstract.php (hunk 1 above)
Comment (johanjanssens) [inline code/filesystem/locator/abstract.php]: You could make this a ===. The result would be the same dirname expects and returns string and so does substr. Or am I missing something ?

File: code/filesystem/locator/component.php (hunk 2 above)
Comment (johanjanssens) [inline code/filesystem/locator/component.php]: Fixed

File: code/filesystem/locator/factory.php (hunk 3 above)
Comment (johanjanssens) [inline code/filesystem/locator/factory.php]: getLocators() doesn't return an associative array. It returns only the keys.

File: code/filesystem/locator/abstract.php (hunk 4 above)
Comment (johanjanssens) [inline code/filesystem/locator/abstract.php]: Not always. The implementation allows you to not specify a format in which case a wildcard will be added and the locator will use the first format it finds.

Comment (oligriffiths) [inline code/filesystem/locator/abstract.php]: Ah, seems that `array_pop()` returns `null` if the array is empty, carry on.

File: code/filesystem/locator/abstract.php (hunk 1 above)
Comment (oligriffiths) [inline code/filesystem/locator/abstract.php]: When comparing strings, `===` is quicker by about 40% as PHP doesn't have to do type coercion checking. I've gotten into the habit of always using `===` for when comparing things I know should always be strings, or the same type in general.

Comment (oligriffiths) [inline code/filesystem/locator/abstract.php]: https://gist.github.com/oligriffiths/4f7b70e2e793f7af2459

Comment (johanjanssens) [inline code/filesystem/locator/abstract.php]: That fair. If we do that we would be best to do a codewise check and optimise everywhere. Micro improvement though.

Comment (oligriffiths) [inline code/filesystem/locator/abstract.php]: Agreed

**Pull request merged**